hypershift-operator/controllers/awsendpointservice/awsendpointservice_controller.go

```diff
@@ -0,0 +1,201 @@
+package awsendpointservice
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/aws/aws-sdk-go/service/elbv2/elbv2iface"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	awsutil "github.com/openshift/hypershift/cmd/infra/aws/util"
+	"github.com/openshift/hypershift/support/upsert"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	finalizer                              = "hypershift.openshift.io/finalizer"
+	endpointServiceDeletionRequeueDuration = time.Duration(5 * time.Second)
+)
+
+type AWSEndpointServiceReconciler struct {
+	client.Client
+	upsert.CreateOrUpdateProvider
+	Region      string
+	ec2Client   ec2iface.EC2API
+	elbv2Client elbv2iface.ELBV2API
+}
+
+func (r *AWSEndpointServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Only add this controller to the manager if we have a non-empty credentials file
+	credsFile := os.Getenv("AWS_SHARED_CREDENTIALS_FILE")
+	if len(credsFile) == 0 {
+		return nil
+	}
+	credsBytes, err := ioutil.ReadFile(credsFile)
+	if err != nil || len(credsBytes) == 0 {
+		return nil
+	}
+
+	_, err = ctrl.NewControllerManagedBy(mgr).
+		For(&hyperv1.AWSEndpointService{}).
+		Build(r)
+	if err != nil {
+		return fmt.Errorf("failed setting up with a controller manager: %w", err)
+	}
+
+	awsSession := awsutil.NewSession("hypershift-operator")
```

Comment:
In https://github.com/openshift/hypershift/pull/636 I am also creating an AWS session elsewhere for the OIDC bucket, do you think we should try to share it or does it not matter?

Reply:
I think we could combine this https://github.com/openshift/hypershift/pull/636/files#diff-9132b8504bc00c929d8a04a3f9235d08a8489ce1effe9fa78eafcc1d7acc7c8fR223-R229

Maybe just do it independently for now and dedup later.